### PR TITLE
Product Query: Add product visibility query support

### LINKED ISSUE #7746: Product Query: query results are different from shortcodes/PHP query loops
## To reproduce

0. Use the sample product data provided by WooCommerce.
1. On a new page, add the Product Query block and `[recent_products]` shortcode block.
2. Change the layout of the Product Query to 4 products per row x 2 rows.
3. Save the page and view it in a new tab.
4. See the products inside the Product Query block and the Recent Products shortcode are different.
5. See the products inside the Product Query block are different between the front end and the editor.
6. See the products inside the editor preview of the Product Query block and the Recent Products shortcode are the same.

## Expected behavior

The query results of the Product Query block should be the same between the editor and front end, as well as the same as the shortcodes/query loops.

## Screenshots

image

### PR DESCRIPTION
**Pull request by dinhtungdu**
Fixes #7746, replaces #7748

This PR adds product visibility support to the Product Query block:
- By default, products that are excluded from the catalog will be hidden.
- When on the search page, products that are excluded from the search will be hidden.
- Stock status query is also updated in favor of product visibility query:
	- When all stock statuses are selected, we return an empty array stock query.
	- When all statuses except `outofstock` is selected, we return an empty stock query and include `outofstock` term to product visibility query.
	- For the remaining set of stock statuses, we return the corresponding meta query.

By using a more complex logic, we can archive the clear UI by selecting all available stock statuses by default, see https://github.com/woocommerce/woocommerce-blocks/pull/7748#issuecomment-1333716474.

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

0. Use the sample product data provided by WooCommerce.
1. On a new page, add the `Producs (Beta)` block and `[recent_products]` shortcode block.
2. Change the layout of the `Producs (Beta)` to 4 products per row x 3 rows.
3. Change the popular preset to `Newest`.
4. Save the page and view it in a new tab.
5. See the products inside the `Producs (Beta)` block (the front end and the editor) and the Recent Products shortcode are the same.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Changelog

> Product Query: Add Product Visibility support.

### CODE REVIEW

**Comment by dinhtungdu** [comment]
I'm fixing the failing unit test. We need to update the test to include product visibility query.

File: src/BlockTypes/ProductQuery.php [hunk 1]
```diff
@@ -325,6 +326,34 @@ function ( $carry, $item ) {
 	 * @return array
 	 */
 	private function get_stock_status_query( $stock_statii ) {
+		$stock_status_options = array_keys( wc_get_product_stock_status_options() );
+
+		if ( ! is_array( $stock_statii ) ) {
+			return array();
+		}
```
Comment (sunyatasattva) [inline src/BlockTypes/ProductQuery.php:331]: Not a big deal, but should we return even earlier? We don't even need a call to `wc_get_product_stock_status_option` if the argument is not an array.

File: src/BlockTypes/ProductQuery.php [hunk 2]
```diff
@@ -325,6 +326,34 @@ function ( $carry, $item ) {
...
+
+		/**
+		 * If all available stock status are selected, we don't need to add the
+		 * meta query for stock status.
+		 */
```
Comment (sunyatasattva) [inline src/BlockTypes/ProductQuery.php:338]: Thanks for this comments! ❤️

**Review by sunyatasattva — APPROVED** [review]
Amazing work! Thanks a lot @dinhtungdu 🌟

File: src/BlockTypes/ProductQuery.php (hunk 1 above)
Comment (dinhtungdu) [reply src/BlockTypes/ProductQuery.php:331]: Absolutely!

**Pull request merged**